piker/data/_buffer.py

```diff
@@ -54,62 +60,92 @@ async def increment_ohlc_buffer(
-                # using ``.push()``?
-
-                # append new entry to buffer thus "incrementing" the bar
-                array = shm.array
-                last = array[-1:][shm._write_fields].copy()
-                # (index, t, close) = last[0][['index', 'time', 'close']]
-                (t, close) = last[0][['time', 'close']]
-
-                # this copies non-std fields (eg. vwap) from the last datum
-                last[
-                    ['time', 'volume', 'open', 'high', 'low', 'close']
-                ][0] = (t + delay_s, 0, close, close, close, close)
-
-                # write to the buffer
-                shm.push(last)
-
-        # broadcast the buffer index step
-        yield {'index': shm._last.value}
-
-
-def subscribe_ohlc_for_increment(
-    shm: ShmArray,
-    delay: int,
+    with trio.CancelScope() as cs:
+
+        # register this time period step as active
+        _incrementers[delay_s] = cs
+        task_status.started(cs)
+
+        while True:
+            # TODO: do we want to support dynamically
+            # adding a "lower" lowest increment period?
+            await trio.sleep(ad)
+            total_s += lowest
+
+            # increment all subscribed shm arrays
+            # TODO: this in ``numba``
+            for delay_s, shms in _shms.items():
+                if total_s % delay_s != 0:
+                    continue
+
+                # TODO: numa this!
+                for shm in shms:
+                    # TODO: in theory we could make this faster by copying the
+                    # "last" readable value into the underlying larger buffer's
+                    # next value and then incrementing the counter instead of
+                    # using ``.push()``?
+
+                    # append new entry to buffer thus "incrementing" the bar
+                    array = shm.array
+                    last = array[-1:][shm._write_fields].copy()
+                    # (index, t, close) = last[0][['index', 'time', 'close']]
+                    (t, close) = last[0][['time', 'close']]
+
+                    # this copies non-std fields (eg. vwap) from the last datum
+                    last[
+                        ['time', 'volume', 'open', 'high', 'low', 'close']
+                    ][0] = (t + delay_s, 0, close, close, close, close)
+
+                    # write to the buffer
+                    shm.push(last)
+
+                # broadcast the buffer index step
+                # yield {'index': shm._last.value}
+                for ctx in _subscribers.get(delay_s, ()):
+                    try:
+                        await ctx.send_yield({'index': shm._last.value})
+                    except (
+                        trio.BrokenResourceError,
+                        trio.ClosedResourceError
+                    ):
+                        log.error(f'{ctx.chan.uid} dropped connection')
+
+
+@tractor.stream
+async def iter_ohlc_periods(
```

Comment:
This is effectively what the simplest dynamic (stream only exists while there is at least one subscriber) pub-sub system looks like.

I'm tempted to entirely drop the stuff in `msg.py` as the subscription API it self seems more confusing than this (`get_topics()`, `topics=List[str]`, and the like).

I think a thorough review and refactor with the `questrade` backend (which has a poll-with-multiple-symbols style producer system)  will probably illuminate what should be discarded.